wrapanapi/systems/google.py

```diff
@@ -0,0 +1,798 @@
+                return data
+            except errors.HttpError as error:
+                if "No such object" in error.content:
+                    self.logger.info(
+                        "File '%s' was not found in bucket '%s'", bucket_name, file_name)
+                else:
+                    raise error
+        return {}
+
+    def upload_file_to_bucket(self, bucket_name, file_path):
+        def handle_progressless_iter(error, progressless_iters):
+            if progressless_iters > NUM_RETRIES:
+                self.logger.info('Failed to make progress for too many consecutive iterations.')
+                raise error
+
+            sleeptime = random.random() * (2 ** progressless_iters)
+            self.logger.info(
+                'Caught exception (%s). Sleeping for %d seconds before retry #%d.',
+                str(error), sleeptime, progressless_iters
+            )
+
+            time.sleep(sleeptime)
+
+        self.logger.info('Building upload request...')
+        media = MediaFileUpload(file_path, chunksize=CHUNKSIZE, resumable=True)
+        if not media.mimetype():
+            media = MediaFileUpload(file_path, DEFAULT_MIMETYPE, resumable=True)
+
+        blob_name = os.path.basename(file_path)
+        if not self.bucket_exists(bucket_name):
+            self.logger.error("Bucket '%s' doesn't exist", bucket_name)
+            raise NotFoundError("bucket {}".format(bucket_name))
+
+        request = self._storage.objects().insert(
+            bucket=bucket_name, name=blob_name, media_body=media)
+        self.logger.info(
+            'Uploading file: %s, to bucket: %s, blob: %s',
+            file_path, bucket_name, blob_name
+        )
+
+        progressless_iters = 0
+        response = None
+        while response is None:
+            error = None
+            try:
+                progress, response = request.next_chunk()
+                if progress:
+                    self.logger.info('Upload progress: %d%%', 100 * progress.progress())
+            except errors.HttpError as error:
+                if error.resp.status < 500:
+                    raise
+            except RETRYABLE_ERRORS as error:
+                if error:
+                    progressless_iters += 1
+                    handle_progressless_iter(error, progressless_iters)
+                else:
+                    progressless_iters = 0
+
+        self.logger.info('Upload complete!')
+        self.logger.info('Uploaded Object:')
+        self.logger.info(json_dumps(response, indent=2))
+        return (True, blob_name)
+
+    def does_forwarding_rule_exist(self, forwarding_rule_name):
+        try:
+            self._find_forwarding_rule_by_name(forwarding_rule_name)
+            return True
+        except errors.HttpError as error:
+            if error.resp.status == 404:
+                return False
+            else:
+                raise
+
+    def list_network(self):
```

Comment:
If we're going to update `list_vm` to be plural, I think the change to plural should be more complete, if possible. That might impact scope of the integration_tests changes, but I think its more confusing to mix the singular and plural method names than to have singular names that return lists (status quo)

Reply:
Yes, exactly, limiting the scope of the PR was the intent here. I planned to leave the unedited methods completely untouched. We can rename them ... it just means we'll have a little more work to do in integration_tests

Reply:
At your discretion then, way to point out all the inconsistencies of the previous implementations ... :P

Reply:
@mshriver I think I am going to hold off on changing the names of all methods so we can limit the scope of this one ... simply renaming all those methods would be a simple (but tedious) PR for several of us to handle soon.